soroban-sdk/src/storage.rs

```diff
@@ -112,52 +137,251 @@ impl Storage {
+        internal::Env::get_contract_data(&self.env, key, storage_type).unwrap_infallible()
+    }
+}
+
+pub struct Unique {
+    env: Env,
+    storage: Storage,
+}
+
+impl Unique {
+    pub fn has<K>(&self, key: &K) -> bool
+    where
+        K: IntoVal<Env, RawVal>,
+    {
+        self.storage.has(key, StorageType::unique())
+    }
+
+    pub fn get<K, V>(&self, key: &K) -> Option<Result<V, V::Error>>
+    where
+        V::Error: Debug,
+        K: IntoVal<Env, RawVal>,
+        V: TryFromVal<Env, RawVal>,
+    {
+        self.storage.get(key, StorageType::unique())
+    }
+
+    pub fn get_unchecked<K, V>(&self, key: &K) -> Result<V, V::Error>
+    where
+        V::Error: Debug,
+        K: IntoVal<Env, RawVal>,
+        V: TryFromVal<Env, RawVal>,
+    {
+        self.storage.get_unchecked(key, StorageType::unique())
+    }
+
+    pub fn set<K, V>(&self, key: &K, val: &V, flags: Option<u32>)
+    where
+        K: IntoVal<Env, RawVal>,
+        V: IntoVal<Env, RawVal>,
+    {
+        self.storage.set(key, val, StorageType::unique(), flags)
+    }
+
+    pub fn bump<K>(&self, key: &K, min_ledgers_to_live: u32)
+    where
+        K: IntoVal<Env, RawVal>,
+    {
+        self.storage
+            .bump(key, StorageType::unique(), min_ledgers_to_live)
+    }
+
+    #[inline(always)]
+    pub fn remove<K>(&self, key: &K)
+    where
+        K: IntoVal<Env, RawVal>,
+    {
+        self.storage.remove(key, StorageType::unique())
+    }
+}
+
+pub struct Recreatable {
+    env: Env,
+    storage: Storage,
+}
+
+impl Recreatable {
+    pub fn has<K>(&self, key: &K) -> bool
+    where
+        K: IntoVal<Env, RawVal>,
+    {
+        self.storage.has(key, StorageType::recreatable())
+    }
+
+    pub fn get<K, V>(&self, key: &K) -> Option<Result<V, V::Error>>
+    where
+        V::Error: Debug,
+        K: IntoVal<Env, RawVal>,
+        V: TryFromVal<Env, RawVal>,
+    {
+        self.storage.get(key, StorageType::recreatable())
+    }
+
+    pub fn get_unchecked<K, V>(&self, key: &K) -> Result<V, V::Error>
+    where
+        V::Error: Debug,
+        K: IntoVal<Env, RawVal>,
+        V: TryFromVal<Env, RawVal>,
+    {
+        self.storage.get_unchecked(key, StorageType::recreatable())
+    }
+
+    pub fn set<K, V>(&self, key: &K, val: &V, flags: Option<u32>)
+    where
+        K: IntoVal<Env, RawVal>,
+        V: IntoVal<Env, RawVal>,
+    {
+        self.storage
+            .set(key, val, StorageType::recreatable(), flags)
+    }
+
+    pub fn bump<K>(&self, key: &K, min_ledgers_to_live: u32)
+    where
+        K: IntoVal<Env, RawVal>,
+    {
+        self.storage
+            .bump(key, StorageType::recreatable(), min_ledgers_to_live)
+    }
+
+    #[inline(always)]
+    pub fn remove<K>(&self, key: &K)
+    where
+        K: IntoVal<Env, RawVal>,
+    {
+        self.storage.remove(key, StorageType::recreatable())
+    }
+}
+
+pub struct Temporary {
+    env: Env,
+    storage: Storage,
```

Comment:
Doesn't look like these sub-structs need `env: Env` in them. The field is unused?

Reply:
removed